File: OnlineTaksiJava/app/src/main/java/com/example/onlinetaksijava/data/remote/socket/SocketManager.java
package com.example.onlinetaksijava.data.remote.socket;

import androidx.annotation.Nullable;

import com.example.onlinetaksijava.util.Constants;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;
import okio.ByteString;

public class SocketManager {

    public interface SocketEventListener {
        void onConnected();
        void onDisconnected();
        void onMessage(String message);
        void onError(String errorMessage);
    }

    private static final OkHttpClient client = new OkHttpClient();
    private static WebSocket webSocket;
    private static SocketEventListener listener;
    private static boolean isConnected = false;

    public static void setListener(@Nullable SocketEventListener newListener) {
        listener = newListener;
    }

    public static void connect(String token) {
        if (isConnected) return;

        Request request = new Request.Builder()
                .url(Constants.WS_BASE_URL + "?token=" + token)
                .build();

        webSocket = client.newWebSocket(request, new WebSocketListener() {
            @Override
            public void onOpen(WebSocket webSocket, Response response) {
                isConnected = true;
                if (listener != null) listener.onConnected();
            }

            @Override
            public void onMessage(WebSocket webSocket, String text) {
                if (listener != null) listener.onMessage(text);
            }

            @Override
            public void onMessage(WebSocket webSocket, ByteString bytes) {
                if (listener != null) listener.onMessage(bytes.utf8());
            }

            @Override
            public void onClosed(WebSocket webSocket, int code, String reason) {
                isConnected = false;
                if (listener != null) listener.onDisconnected();
            }

            @Override
            public void onClosing(WebSocket webSocket, int code, String reason) {
                isConnected = false;
                if (listener != null) listener.onDisconnected();
            }

            @Override
            public void onFailure(WebSocket webSocket, Throwable t, Response response) {
                isConnected = false;
                if (listener != null) listener.onError(t.getMessage() != null ? t.getMessage() : "Socket hatası");
            }
        });
    }

    public static void sendRawJson(String jsonText) {
        if (webSocket != null) {
            webSocket.send(jsonText);
        }
    }

    public static void sendPing() {
        sendRawJson("{\"event\":\"PING\",\"data\":{}}");
    }

    public static void disconnect() {
        isConnected = false;
        if (webSocket != null) {
            webSocket.close(1000, "Client disconnected");
            webSocket = null;
        }
    }

    public static boolean isSocketConnected() {
        return isConnected;
    }
}

File: OnlineTaksiJetpack/app/src/main/java/com/example/onlinetaksijetpack/data/remote/socket/SocketManager.kt
package com.example.onlinetaksijetpack.data.remote.socket


import com.example.onlinetaksijetpack.util.Constants
import okhttp3.*
import okio.ByteString

object SocketManager {

    interface SocketEventListener {
        fun onConnected()
        fun onDisconnected()
        fun onMessage(message: String)
        fun onError(errorMessage: String)
    }

    private val client = OkHttpClient()

    private var webSocket: WebSocket? = null
    private var listener: SocketEventListener? = null
    private var isConnected = false

    fun setListener(newListener: SocketEventListener?) {
        listener = newListener
    }

    fun connect(token: String) {
        if (isConnected) return

        val request = Request.Builder()
            .url("${Constants.WS_BASE_URL}?token=$token")
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                isConnected = true
                listener?.onConnected()
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                listener?.onMessage(text)
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                listener?.onMessage(bytes.utf8())
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                isConnected = false
                listener?.onDisconnected()
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                isConnected = false
                listener?.onDisconnected()
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                isConnected = false
                listener?.onError(t.message ?: "Socket hatası")
            }
        })
    }

    fun sendPing() {
        webSocket?.send("""{"event":"PING","data":{}}""")
    }

    fun disconnect() {
        isConnected = false
        webSocket?.close(1000, "Client disconnected")
        webSocket = null
    }

    fun isSocketConnected(): Boolean = isConnected
}